iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_distributor.inl

```diff
@@ -132,23 +132,67 @@ inline bool ChunkDistributor<ChunkDistributorDataType>::hasStoredQueues() const
 template <typename ChunkDistributorDataType>
 inline void ChunkDistributor<ChunkDistributorDataType>::deliverToAllStoredQueues(mepoo::SharedChunk chunk) noexcept
 {
-    typename MemberType_t::LockGuard_t lock(*getMembers());
+    typename ChunkDistributorDataType::QueueContainer_t remainingQueues;
+    {
+        typename MemberType_t::LockGuard_t lock(*getMembers());
+
+        // send to all the queues
+        for (auto& queue : getMembers()->m_queues)
+        {
+            bool isBlockingQueue =
+                (getMembers()->m_subscriberTooSlowPolicy == SubscriberTooSlowPolicy::WAIT_FOR_SUBSCRIBER
+                 && queue->m_queueFullPolicy == QueueFullPolicy::BLOCK_PUBLISHER);
 
-    // send to all the queues
-    for (auto& queue : getMembers()->m_queues)
+            if (!deliverToQueue(queue.get(), chunk) && isBlockingQueue)
+            {
+                remainingQueues.emplace_back(queue);
+            }
+        }
+    }
+
+    // busy waiting until every queue is served
+    while (!remainingQueues.empty())
     {
-        deliverToQueue(queue.get(), chunk);
+        std::this_thread::yield();
+        {
+            // create intersection of current queues and remainingQueues
+            typename MemberType_t::LockGuard_t lock(*getMembers());
+            typename ChunkDistributorDataType::QueueContainer_t queueIntersection(remainingQueues.size());
+            std::sort(getMembers()->m_queues.begin(), getMembers()->m_queues.end());
+            std::sort(remainingQueues.begin(), remainingQueues.end());
+
+            auto iter = std::set_intersection(getMembers()->m_queues.begin(),
+                                              getMembers()->m_queues.end(),
+                                              remainingQueues.begin(),
+                                              remainingQueues.end(),
+                                              queueIntersection.begin());
+            queueIntersection.resize(static_cast<uint64_t>(iter - queueIntersection.begin()));
+            remainingQueues = queueIntersection;
+
+            // deliver to remaining queues
+            for (uint64_t i = remainingQueues.size() - 1U; !remainingQueues.empty(); --i)
+            {
+                if (deliverToQueue(remainingQueues[i].get(), chunk))
+                {
+                    remainingQueues.erase(remainingQueues.begin() + i);
+                }
+
+                if (i == 0U)
+                {
+                    break;
+                }
```

Comment:
This should be the very first check aka early exit.

Reply:
Nope then the algorithm would break since I am counting backwards. An overflow checks is present see `!remainingQueus.empty()`